xds/internal/balancer/ringhash/e2e/ringhash_balancer_test.go

```diff
@@ -134,11 +142,318 @@ func (s) TestRingHash_ReconnectToMoveOutOfTransientFailure(t *testing.T) {
 		}
 	}
 	if err := ctx.Err(); err != nil {
-		t.Fatalf("Timeout waiting for channel to reach READT after server restart: %v", err)
+		t.Fatalf("Timeout waiting for channel to reach READY after server restart: %v", err)
 	}
 
 	// An RPC at this point is expected to fail.
 	if _, err := client.EmptyCall(ctx, &testpb.Empty{}); err != nil {
 		t.Fatalf("rpc EmptyCall() failed: %v", err)
 	}
 }
+
+// XXX this is copy pasted from
+// https://github.com/grpc/grpc-go/blob/9199290ff8fee7c37283586929af39ba23dbab4b/xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go#L93
+// Perhaps we should move it to a common package?
+func startTestServiceBackends(t *testing.T, numBackends int) ([]*stubserver.StubServer, func()) {
+	t.Helper()
+	var servers []*stubserver.StubServer
+	for i := 0; i < numBackends; i++ {
+		servers = append(servers, stubserver.StartTestService(t, &stubserver.StubServer{
+			UnaryCallF: func(ctx context.Context, _ *testpb.SimpleRequest) (*testpb.SimpleResponse, error) {
+				return &testpb.SimpleResponse{
+					Hostname: strconv.Itoa(i),
+				}, nil
+			},
+		}))
+	}
+
+	return servers, func() {
+		for _, server := range servers {
+			server.Stop()
+		}
+	}
+}
+
+// makeNonExistingBackends starts numBackends backends and stops them
+// immediately to simulate a server that is not reachable.
+func makeNonExistingBackends(t *testing.T, numBackends int) []*stubserver.StubServer {
+	var servers []*stubserver.StubServer
+	for i := 0; i < numBackends; i++ {
+		servers = append(servers, stubserver.StartTestService(t, nil))
+		servers[i].Stop()
```

Comment:
Can this lead to test flakes at all, if the port released here (after the server is closed) is used by another test backend spun up by the same test binary?

Reply:
Put another way, is there a different/better way to simulate a server that is not reachable?

Reply:
Yeah, we don't run tests in parallel. That would require us to move some mountains (or maybe not, maybe its just that our leak checker doesn't like parallel tests).